server/src/main/java/org/opensearch/action/admin/cluster/remotestore/restore/TransportRestoreRemoteStoreAction.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.action.admin.cluster.remotestore.restore;
+
+import org.opensearch.action.ActionListener;
+import org.opensearch.action.support.ActionFilters;
+import org.opensearch.action.support.clustermanager.TransportClusterManagerNodeAction;
+import org.opensearch.cluster.ClusterState;
+import org.opensearch.cluster.block.ClusterBlockException;
+import org.opensearch.cluster.block.ClusterBlockLevel;
+import org.opensearch.cluster.metadata.IndexNameExpressionResolver;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.inject.Inject;
+import org.opensearch.common.io.stream.StreamInput;
+import org.opensearch.snapshots.RestoreInfo;
+import org.opensearch.snapshots.RestoreService;
+import org.opensearch.threadpool.ThreadPool;
+import org.opensearch.transport.TransportService;
+
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * Transport action for restore remote store operation
+ *
+ * @opensearch.internal
+ */
+public class TransportRestoreRemoteStoreAction extends TransportClusterManagerNodeAction<
+    RestoreRemoteStoreRequest,
+    RestoreRemoteStoreResponse> {
+    private final RestoreService restoreService;
+
+    @Inject
+    public TransportRestoreRemoteStoreAction(
+        TransportService transportService,
+        ClusterService clusterService,
+        ThreadPool threadPool,
+        RestoreService restoreService,
+        ActionFilters actionFilters,
+        IndexNameExpressionResolver indexNameExpressionResolver
+    ) {
+        super(
+            RestoreRemoteStoreAction.NAME,
+            transportService,
+            clusterService,
+            threadPool,
+            actionFilters,
+            RestoreRemoteStoreRequest::new,
+            indexNameExpressionResolver
+        );
+        this.restoreService = restoreService;
+    }
+
+    @Override
+    protected String executor() {
+        return ThreadPool.Names.GENERIC;
+    }
```

Comment:
We need to decide on the threadpool. I'd prefer having a dedicated threadpool

Reply:
Let me think a bit more on this. Will get back.

Reply:
This is on the same lines as snapshot restore (https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/action/admin/cluster/snapshots/restore/TransportRestoreSnapshotAction.java#L82). 

Currently, we don't have a dedicated threadpool for remote store as the segment upload happens in the refresh flow.

Reply:
@Bukhtawar Does it answer your question?

Reply:
We need to think about what happens to segments when remote store isn't available. Do we block flushes eventually?

Reply:
If remote store is not available, we will not block flush directly. If segment upload is not working, we will not purge translog. If translog grows beyond some threshold, we will block write operations. This will indirectly block flush.

> Also need to specify if this can run concurrently with a snapshot segment restore

At a time, only one restore operation will happen. In the implementation of the API (https://github.com/opensearch-project/OpenSearch/pull/3642), we change the state of the index to `INITIALIZING`, if index state is already INITIALIZING, we fail the operation.

Reply:
I agree with @Bukhtawar , we need to use `SNAPSHOT` (why `TransportRestoreSnapshotAction` is not using it?). thread pool or introduce a dedicated one ...